src/main/java/org/candlepin/pki/impl/BouncyCastlePKIUtility.java

```diff
@@ -0,0 +1,650 @@
+
+        KeyFactory factory = KeyFactory.getInstance(algorithm, BouncyCastleProviderLoader.getProvider());
+        PKCS8EncodedKeySpec spec = new PKCS8EncodedKeySpec(keydata, algorithm);
+
+        return factory.generatePrivate(spec);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public KeyPair generateKeyPair() throws KeyException {
+        try {
+            KeyPairGenerator keyGen = KeyPairGenerator.getInstance(KEY_ALGORITHM);
+            keyGen.initialize(KEY_SIZE);
+            KeyPair keypair = keyGen.generateKeyPair();
+            return this.buildInsecureKeyPair(keypair);
+        }
+        catch (NoSuchAlgorithmException e) {
+            throw new KeyException(e);
+        }
+    }
+
+    /**
+     * Rebuilds an "insecure" KeyPair instance from another KeyPair instance. The returned KeyPair
+     * will allow fetching the encoded form of both the public and private key.
+     *
+     * @param keypair
+     *  the keypair to rebuild
+     *
+     * @return
+     *  a keypair which allows extracting the encoded form of both the public and private key
+     */
+    private KeyPair buildInsecureKeyPair(KeyPair keypair) throws KeyException {
+        return this.buildInsecureKeyPair(keypair.getPublic(), keypair.getPrivate());
+    }
+
+    /**
+     * Builds an "insecure" KeyPair instance from the provided public and private keys. The returned
+     * KeyPair will allow fetching the encoded form of both the public and private key.
+     *
+     * @param publicKey
+     *  the public key to use to build the insecure keypair
+     *
+     * @param privateKey
+     *  the private key to use to build the insecure keypair
+     *
+     * @return
+     *  a keypair which allows extracting the encoded form of both the public and private key
+     */
+    private KeyPair buildInsecureKeyPair(PublicKey publicKey, PrivateKey privateKey)
+        throws KeyException {
+
+        byte[] pubEncoded = publicKey.getEncoded();
+        byte[] privEncoded = this.getKeyEncoding(privateKey);
+
+        PublicKey insecurePub = new BouncyCastlePKIUtility.InsecurePublicKey(publicKey, pubEncoded,
+            publicKey.getFormat());
+        PrivateKey insecurePriv = new BouncyCastlePKIUtility.InsecurePrivateKey(privateKey, privEncoded,
+            "PKCS#8");
+
+        return new KeyPair(insecurePub, insecurePriv);
+    }
+
+    /**
+     * Fetches the encoded form of the specified private key
+     *
+     * @param privateKey
+     *  the private key from which to fetch the encoded form
+     *
+     * @return
+     *  the encoded form of the given private key
+     */
+    private byte[] getKeyEncoding(PrivateKey privateKey) throws KeyException {
+        byte[] unwrapped = privateKey.getEncoded();
+
+        if (unwrapped != null) {
+            return unwrapped;
+        }
+
+        try {
+            String algorithm = "AES";
+            String transformation = "AES/CBC/PKCS5Padding";
```

Comment:
We can replace this with "AES/GCM/NoPadding" to resolve the sonarcloud vulnerability alerts.
However, it needs to be tested with subscription-manager (registration, entitlement cert generation, etc.) to make sure this is compatible.

Reply:
This is the same transformation JSS was using. How did we get past it before? Is it because sonarcloud is only monitoring changes?

Reply:
is subman good with this change?